exercises/Robert_Applin/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,203 @@
+                     std::string::const_iterator toIterator,
+                     const std::string &characters) {
+  for (auto it = fromIterator; it < toIterator; ++it) {
+    if (characters.find(*it) == std::string::npos)
+      return it;
+  }
+  return toIterator;
+}
+
+/**
+ * @return A vector of sub strings split around delimiters
+ */
+std::vector<std::string> splitBy(std::string::const_iterator fromIterator,
+                                 std::string::const_iterator toIterator,
+                                 const std::string &delimiters) {
+  std::vector<std::string> substrings;
+  auto iter = findNextNotCharacter(fromIterator, toIterator, delimiters);
+
+  while (iter != toIterator) {
+    const auto nextIter = findNextCharacter(iter, toIterator, delimiters);
+    substrings.emplace_back(iter, nextIter);
+    iter = findNextNotCharacter(nextIter, toIterator, delimiters);
+  }
+  return substrings;
+}
+
+/**
+ * @return A true if count1 is larger than count2
+ */
+bool compareCounts(const std::pair<std::string, std::size_t> &count1,
+                   const std::pair<std::string, std::size_t> &count2) {
+  return count1.second > count2.second;
+}
+
+/**
+ * @return True if the string is shorter than 5 characters
+ */
+bool isShorterThanFive(const std::string &str) { return str.length() < 5; }
+
+/**
+ * Removes any words shorter than five from the vector
+ */
+void removeShorterThanFive(std::vector<std::string> &strings) {
+  auto it = std::remove_if(strings.begin(), strings.end(), isShorterThanFive);
+  // Remove invalid indices
+  strings.erase(it, strings.end());
+}
+
+/**
+ * @return A string transformed to lower case
+ */
+std::string toLower(std::string str) {
+  std::transform(str.begin(), str.end(), str.begin(),
+                 static_cast<int (*)(int)>(std::tolower));
+  return str;
+}
+
+/**
+ * Add one to the count for a string, or creates a new mapping for a
+ * new string and sets its count to one.
+ */
+void addToCount(std::unordered_map<std::string, std::size_t> &count,
+                const std::string &str) {
+  auto it = count.find(str);
+  if (it == count.end())
+    count[str] = 1;
+  else
+    ++it->second;
+}
+
+/**
+ * Loops through the strings in substrings and records the
+ * number of occurences for each string.
+ */
+void countWords(const std::vector<std::string> &substrings,
+                std::unordered_map<std::string, std::size_t> &count) {
+  for (auto const &str : substrings)
+    addToCount(count, str);
+}
+
+/**
+ * Splits the string into each word, removes short words and changes
+ * each word to lower case before counting the number of occurances.
+ */
+void countWords(const std::string &str, const std::string &delimiters,
+                std::unordered_map<std::string, std::size_t> &count) {
+  auto words = splitBy(str.begin(), str.end(), delimiters);
+  removeShorterThanFive(words);
+  std::transform(words.begin(), words.end(), words.begin(), toLower);
```

Comment:
You wrote a `toLower` function, but you haven't used it here.

Reply:
I think I have used it there

Reply:
Yes but I think you could have done:
```cpp
words = toLower(words)
```

?

Reply:
Yes, you are right!